docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+# Design and Coding Guidelines
+
+## Design Phase
+
+For non-trivial changes, please communicate your plans early on to reach maximum impact and to avoid duplicate work:
+
+* What are you planning to working on?
+* Why is it needed for the project?
+* Why is it needed in the repository?
+* Ask the rest of the team if they have already worked on it or thought about it. This helps to get to know about existing code fragments that can simplify your task, and integrate better with existing code.
```

Comment:
This assumes the contributor is in the team.

Reply:
I've added some more phrasing, but please add your thoughts and suggestions here.

Reply:
I guess my comment about discussions/issues above applies here too.